// @scrypted/sdk/dist/types/gen/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScryptedMimeTypes = exports.ScryptedInterface = exports.MediaPlayerState = exports.SecuritySystemObstruction = exports.SecuritySystemMode = exports.AirQuality = exports.AirPurifierMode = exports.AirPurifierStatus = exports.ChargeState = exports.LockState = exports.PanTiltZoomMovement = exports.ThermostatMode = exports.TemperatureUnit = exports.FanMode = exports.HumidityMode = exports.ScryptedDeviceType = exports.ScryptedInterfaceDescriptors = exports.ScryptedInterfaceMethod = exports.ScryptedInterfaceProperty = exports.DeviceBase = exports.TYPES_VERSION = void 0;
exports.TYPES_VERSION = "0.2.98";
class DeviceBase {
}
exports.DeviceBase = DeviceBase;
var ScryptedInterfaceProperty;
(function (ScryptedInterfaceProperty) {
    ScryptedInterfaceProperty["id"] = "id";
    ScryptedInterfaceProperty["info"] = "info";
    ScryptedInterfaceProperty["interfaces"] = "interfaces";
    ScryptedInterfaceProperty["mixins"] = "mixins";
    ScryptedInterfaceProperty["name"] = "name";
    ScryptedInterfaceProperty["nativeId"] = "nativeId";
    ScryptedInterfaceProperty["pluginId"] = "pluginId";
    ScryptedInterfaceProperty["providedInterfaces"] = "providedInterfaces";
    ScryptedInterfaceProperty["providedName"] = "providedName";
    ScryptedInterfaceProperty["providedRoom"] = "providedRoom";
    ScryptedInterfaceProperty["providedType"] = "providedType";
    ScryptedInterfaceProperty["providerId"] = "providerId";
    ScryptedInterfaceProperty["room"] = "room";
    ScryptedInterfaceProperty["type"] = "type";
    ScryptedInterfaceProperty["on"] = "on";
    ScryptedInterfaceProperty["brightness"] = "brightness";
    ScryptedInterfaceProperty["colorTemperature"] = "colorTemperature";
    ScryptedInterfaceProperty["rgb"] = "rgb";
    ScryptedInterfaceProperty["hsv"] = "hsv";
    ScryptedInterfaceProperty["running"] = "running";
    ScryptedInterfaceProperty["paused"] = "paused";
    ScryptedInterfaceProperty["docked"] = "docked";
    ScryptedInterfaceProperty["temperatureSetting"] = "temperatureSetting";
    ScryptedInterfaceProperty["thermostatActiveMode"] = "thermostatActiveMode";
    ScryptedInterfaceProperty["thermostatAvailableModes"] = "thermostatAvailableModes";
    ScryptedInterfaceProperty["thermostatMode"] = "thermostatMode";
    ScryptedInterfaceProperty["thermostatSetpoint"] = "thermostatSetpoint";
    ScryptedInterfaceProperty["thermostatSetpointHigh"] = "thermostatSetpointHigh";
    ScryptedInterfaceProperty["thermostatSetpointLow"] = "thermostatSetpointLow";
    ScryptedInterfaceProperty["temperature"] = "temperature";
    ScryptedInterfaceProperty["temperatureUnit"] = "temperatureUnit";
    ScryptedInterfaceProperty["humidity"] = "humidity";
    ScryptedInterfaceProperty["recordingActive"] = "recordingActive";
    ScryptedInterfaceProperty["ptzCapabilities"] = "ptzCapabilities";
    ScryptedInterfaceProperty["lockState"] = "lockState";
    ScryptedInterfaceProperty["entryOpen"] = "entryOpen";
    ScryptedInterfaceProperty["batteryLevel"] = "batteryLevel";
    ScryptedInterfaceProperty["chargeState"] = "chargeState";
    ScryptedInterfaceProperty["online"] = "online";
    ScryptedInterfaceProperty["fromMimeType"] = "fromMimeType";
    ScryptedInterfaceProperty["toMimeType"] = "toMimeType";
    ScryptedInterfaceProperty["binaryState"] = "binaryState";
    ScryptedInterfaceProperty["tampered"] = "tampered";
    ScryptedInterfaceProperty["powerDetected"] = "powerDetected";
    ScryptedInterfaceProperty["audioDetected"] = "audioDetected";
    ScryptedInterfaceProperty["motionDetected"] = "motionDetected";
    ScryptedInterfaceProperty["ambientLight"] = "ambientLight";
    ScryptedInterfaceProperty["occupied"] = "occupied";
    ScryptedInterfaceProperty["flooded"] = "flooded";
    ScryptedInterfaceProperty["ultraviolet"] = "ultraviolet";
    ScryptedInterfaceProperty["luminance"] = "luminance";
    ScryptedInterfaceProperty["position"] = "position";
    ScryptedInterfaceProperty["securitySystemState"] = "securitySystemState";
    ScryptedInterfaceProperty["pm10Density"] = "pm10Density";
    ScryptedInterfaceProperty["pm25Density"] = "pm25Density";
    ScryptedInterfaceProperty["vocDensity"] = "vocDensity";
    ScryptedInterfaceProperty["noxDensity"] = "noxDensity";
    ScryptedInterfaceProperty["co2ppm"] = "co2ppm";
    ScryptedInterfaceProperty["airQuality"] = "airQuality";
    ScryptedInterfaceProperty["airPurifierState"] = "airPurifierState";
    ScryptedInterfaceProperty["filterChangeIndication"] = "filterChangeIndication";
    ScryptedInterfaceProperty["filterLifeLevel"] = "filterLifeLevel";
    ScryptedInterfaceProperty["humiditySetting"] = "humiditySetting";
    ScryptedInterfaceProperty["fan"] = "fan";
    ScryptedInterfaceProperty["applicationInfo"] = "applicationInfo";
})(ScryptedInterfaceProperty = exports.ScryptedInterfaceProperty || (exports.ScryptedInterfaceProperty = {}));
var ScryptedInterfaceMethod;
(function (ScryptedInterfaceMethod) {
    ScryptedInterfaceMethod["listen"] = "listen";
    ScryptedInterfaceMethod["probe"] = "probe";
    ScryptedInterfaceMethod["setMixins"] = "setMixins";
    ScryptedInterfaceMethod["setName"] = "setName";
    ScryptedInterfaceMethod["setRoom"] = "setRoom";
    ScryptedInterfaceMethod["setType"] = "setType";
    ScryptedInterfaceMethod["getPluginJson"] = "getPluginJson";
    ScryptedInterfaceMethod["turnOff"] = "turnOff";
    ScryptedInterfaceMethod["turnOn"] = "turnOn";
    ScryptedInterfaceMethod["setBrightness"] = "setBrightness";
    ScryptedInterfaceMethod["getTemperatureMaxK"] = "getTemperatureMaxK";
    ScryptedInterfaceMethod["getTemperatureMinK"] = "getTemperatureMinK";
    ScryptedInterfaceMethod["setColorTemperature"] = "setColorTemperature";
    ScryptedInterfaceMethod["setRgb"] = "setRgb";
    ScryptedInterfaceMethod["setHsv"] = "setHsv";
    ScryptedInterfaceMethod["sendNotification"] = "sendNotification";
    ScryptedInterfaceMethod["start"] = "start";
    ScryptedInterfaceMethod["stop"] = "stop";
    ScryptedInterfaceMethod["pause"] = "pause";
    ScryptedInterfaceMethod["resume"] = "resume";
    ScryptedInterfaceMethod["dock"] = "dock";
    ScryptedInterfaceMethod["setTemperature"] = "setTemperature";
    ScryptedInterfaceMethod["setThermostatMode"] = "setThermostatMode";
    ScryptedInterfaceMethod["setThermostatSetpoint"] = "setThermostatSetpoint";
    ScryptedInterfaceMethod["setThermostatSetpointHigh"] = "setThermostatSetpointHigh";
    ScryptedInterfaceMethod["setThermostatSetpointLow"] = "setThermostatSetpointLow";
    ScryptedInterfaceMethod["setTemperatureUnit"] = "setTemperatureUnit";
    ScryptedInterfaceMethod["getPictureOptions"] = "getPictureOptions";
    ScryptedInterfaceMethod["takePicture"] = "takePicture";
    ScryptedInterfaceMethod["getAudioStream"] = "getAudioStream";
    ScryptedInterfaceMethod["startDisplay"] = "startDisplay";
    ScryptedInterfaceMethod["stopDisplay"] = "stopDisplay";
    ScryptedInterfaceMethod["getVideoStream"] = "getVideoStream";
    ScryptedInterfaceMethod["getVideoStreamOptions"] = "getVideoStreamOptions";
    ScryptedInterfaceMethod["getRecordingStream"] = "getRecordingStream";
    ScryptedInterfaceMethod["getRecordingStreamCurrentTime"] = "getRecordingStreamCurrentTime";
    ScryptedInterfaceMethod["getRecordingStreamOptions"] = "getRecordingStreamOptions";
    ScryptedInterfaceMethod["getRecordingStreamThumbnail"] = "getRecordingStreamThumbnail";
    ScryptedInterfaceMethod["deleteRecordingStream"] = "deleteRecordingStream";
    ScryptedInterfaceMethod["setRecordingActive"] = "setRecordingActive";
    ScryptedInterfaceMethod["ptzCommand"] = "ptzCommand";
    ScryptedInterfaceMethod["getRecordedEvents"] = "getRecordedEvents";
    ScryptedInterfaceMethod["getVideoClip"] = "getVideoClip";
    ScryptedInterfaceMethod["getVideoClipThumbnail"] = "getVideoClipThumbnail";
    ScryptedInterfaceMethod["getVideoClips"] = "getVideoClips";
    ScryptedInterfaceMethod["removeVideoClips"] = "removeVideoClips";
    ScryptedInterfaceMethod["setVideoStreamOptions"] = "setVideoStreamOptions";
    ScryptedInterfaceMethod["startIntercom"] = "startIntercom";
    ScryptedInterfaceMethod["stopIntercom"] = "stopIntercom";
    ScryptedInterfaceMethod["lock"] = "lock";
    ScryptedInterfaceMethod["unlock"] = "unlock";
    ScryptedInterfaceMethod["addPassword"] = "addPassword";
    ScryptedInterfaceMethod["getPasswords"] = "getPasswords";
    ScryptedInterfaceMethod["removePassword"] = "removePassword";
    ScryptedInterfaceMethod["activate"] = "activate";
    ScryptedInterfaceMethod["deactivate"] = "deactivate";
    ScryptedInterfaceMethod["isReversible"] = "isReversible";
    ScryptedInterfaceMethod["closeEntry"] = "closeEntry";
    ScryptedInterfaceMethod["openEntry"] = "openEntry";
    ScryptedInterfaceMethod["getDevice"] = "getDevice";
    ScryptedInterfaceMethod["releaseDevice"] = "releaseDevice";
    ScryptedInterfaceMethod["adoptDevice"] = "adoptDevice";
    ScryptedInterfaceMethod["discoverDevices"] = "discoverDevices";
    ScryptedInterfaceMethod["createDevice"] = "createDevice";
    ScryptedInterfaceMethod["getCreateDeviceSettings"] = "getCreateDeviceSettings";
    ScryptedInterfaceMethod["reboot"] = "reboot";
    ScryptedInterfaceMethod["getRefreshFrequency"] = "getRefreshFrequency";
    ScryptedInterfaceMethod["refresh"] = "refresh";
    ScryptedInterfaceMethod["getMediaStatus"] = "getMediaStatus";
    ScryptedInterfaceMethod["load"] = "load";
    ScryptedInterfaceMethod["seek"] = "seek";
    ScryptedInterfaceMethod["skipNext"] = "skipNext";
    ScryptedInterfaceMethod["skipPrevious"] = "skipPrevious";
    ScryptedInterfaceMethod["convert"] = "convert";
    ScryptedInterfaceMethod["getSettings"] = "getSettings";
    ScryptedInterfaceMethod["putSetting"] = "putSetting";
    ScryptedInterfaceMethod["armSecuritySystem"] = "armSecuritySystem";
    ScryptedInterfaceMethod["disarmSecuritySystem"] = "disarmSecuritySystem";
    ScryptedInterfaceMethod["setAirPurifierState"] = "setAirPurifierState";
    ScryptedInterfaceMethod["getReadmeMarkdown"] = "getReadmeMarkdown";
    ScryptedInterfaceMethod["getOauthUrl"] = "getOauthUrl";
    ScryptedInterfaceMethod["onOauthCallback"] = "onOauthCallback";
    ScryptedInterfaceMethod["canMixin"] = "canMixin";
    ScryptedInterfaceMethod["getMixin"] = "getMixin";
    ScryptedInterfaceMethod["releaseMixin"] = "releaseMixin";
    ScryptedInterfaceMethod["onRequest"] = "onRequest";
    ScryptedInterfaceMethod["onConnection"] = "onConnection";
    ScryptedInterfaceMethod["onPush"] = "onPush";
    ScryptedInterfaceMethod["run"] = "run";
    ScryptedInterfaceMethod["eval"] = "eval";
    ScryptedInterfaceMethod["loadScripts"] = "loadScripts";
    ScryptedInterfaceMethod["saveScript"] = "saveScript";
    ScryptedInterfaceMethod["trackObjects"] = "trackObjects";
    ScryptedInterfaceMethod["getDetectionInput"] = "getDetectionInput";
    ScryptedInterfaceMethod["getObjectTypes"] = "getObjectTypes";
    ScryptedInterfaceMethod["detectObjects"] = "detectObjects";
    ScryptedInterfaceMethod["generateObjectDetections"] = "generateObjectDetections";
    ScryptedInterfaceMethod["getDetectionModel"] = "getDetectionModel";
    ScryptedInterfaceMethod["setHumidity"] = "setHumidity";
    ScryptedInterfaceMethod["setFan"] = "setFan";
    ScryptedInterfaceMethod["startRTCSignalingSession"] = "startRTCSignalingSession";
    ScryptedInterfaceMethod["createRTCSignalingSession"] = "createRTCSignalingSession";
    ScryptedInterfaceMethod["getScryptedUserAccessControl"] = "getScryptedUserAccessControl";
    ScryptedInterfaceMethod["generateVideoFrames"] = "generateVideoFrames";
    ScryptedInterfaceMethod["connectStream"] = "connectStream";
})(ScryptedInterfaceMethod = exports.ScryptedInterfaceMethod || (exports.ScryptedInterfaceMethod = {}));
exports.ScryptedInterfaceDescriptors = {
    ScryptedDevice: {
        name: 'ScryptedDevice',
        methods: [
            'listen',
            'probe',
            'setMixins',
            'setName',
            'setRoom',
            'setType'
        ],
        properties: [
            'id',
            'info',
            'interfaces',
            'mixins',
            'name',
            'nativeId',
            'pluginId',
            'providedInterfaces',
            'providedName',
            'providedRoom',
            'providedType',
            'providerId',
            'room',
            'type'
        ]
    },
    ScryptedPlugin: {
        name: 'ScryptedPlugin',
        methods: [
            'getPluginJson'
        ],
        properties: []
    },
    OnOff: {
        name: 'OnOff',
        methods: [
            'turnOff',
            'turnOn'
        ],
        properties: [
            'on'
        ]
    },
    Brightness: {
        name: 'Brightness',
        methods: [
            'setBrightness'
        ],
        properties: [
            'brightness'
        ]
    },
    ColorSettingTemperature: {
        name: 'ColorSettingTemperature',
        methods: [
            'getTemperatureMaxK',
            'getTemperatureMinK',
            'setColorTemperature'
        ],
        properties: [
            'colorTemperature'
        ]
    },
    ColorSettingRgb: {
        name: 'ColorSettingRgb',
        methods: [
            'setRgb'
        ],
        properties: [
            'rgb'
        ]
    },
    ColorSettingHsv: {
        name: 'ColorSettingHsv',
        methods: [
            'setHsv'
        ],
        properties: [
            'hsv'
        ]
    },
    Notifier: {
        name: 'Notifier',
        methods: [
            'sendNotification'
        ],
        properties: []
    },
    StartStop: {
        name: 'StartStop',
        methods: [
            'start',
            'stop'
        ],
        properties: [
            'running'
        ]
    },
    Pause: {
        name: 'Pause',
        methods: [
            'pause',
            'resume'
        ],
        properties: [
            'paused'
        ]
    },
    Dock: {
        name: 'Dock',
        methods: [
            'dock'
        ],
        properties: [
            'docked'
        ]
    },
    TemperatureSetting: {
        name: 'TemperatureSetting',
        methods: [
            'setTemperature',
            'setThermostatMode',
            'setThermostatSetpoint',
            'setThermostatSetpointHigh',
            'setThermostatSetpointLow'
        ],
        properties: [
            'temperatureSetting',
            'thermostatActiveMode',
            'thermostatAvailableModes',
            'thermostatMode',
            'thermostatSetpoint',
            'thermostatSetpointHigh',
            'thermostatSetpointLow'
        ]
    },
    Thermometer: {
        name: 'Thermometer',
        methods: [
            'setTemperatureUnit'
        ],
        properties: [
            'temperature',
            'temperatureUnit'
        ]
    },
    HumiditySensor: {
        name: 'HumiditySensor',
        methods: [],
        properties: [
            'humidity'
        ]
    },
    Camera: {
        name: 'Camera',
        methods: [
            'getPictureOptions',
            'takePicture'
        ],
        properties: []
    },
    Microphone: {
        name: 'Microphone',
        methods: [
            'getAudioStream'
        ],
        properties: []
    },
    Display: {
        name: 'Display',
        methods: [
            'startDisplay',
            'stopDisplay'
        ],
        properties: []
    },
    VideoCamera: {
        name: 'VideoCamera',
        methods: [
            'getVideoStream',
            'getVideoStreamOptions'
        ],
        properties: []
    },
    VideoRecorder: {
        name: 'VideoRecorder',
        methods: [
            'getRecordingStream',
            'getRecordingStreamCurrentTime',
            'getRecordingStreamOptions',
            'getRecordingStreamThumbnail'
        ],
        properties: [
            'recordingActive'
        ]
    },
    VideoRecorderManagement: {
        name: 'VideoRecorderManagement',
        methods: [
            'deleteRecordingStream',
            'setRecordingActive'
        ],
        properties: []
    },
    PanTiltZoom: {
        name: 'PanTiltZoom',
        methods: [
            'ptzCommand'
        ],
        properties: [
            'ptzCapabilities'
        ]
    },
    EventRecorder: {
        name: 'EventRecorder',
        methods: [
            'getRecordedEvents'
        ],
        properties: []
    },
    VideoClips: {
        name: 'VideoClips',
        methods: [
            'getVideoClip',
            'getVideoClipThumbnail',
            'getVideoClips',
            'removeVideoClips'
        ],
        properties: []
    },
    VideoCameraConfiguration: {
        name: 'VideoCameraConfiguration',
        methods: [
            'setVideoStreamOptions'
        ],
        properties: []
    },
    Intercom: {
        name: 'Intercom',
        methods: [
            'startIntercom',
            'stopIntercom'
        ],
        properties: []
    },
    Lock: {
        name: 'Lock',
        methods: [
            'lock',
            'unlock'
        ],
        properties: [
            'lockState'
        ]
    },
    PasswordStore: {
        name: 'PasswordStore',
        methods: [
            'addPassword',
            'getPasswords',
            'removePassword'
        ],
        properties: []
    },
    Scene: {
        name: 'Scene',
        methods: [
            'activate',
            'deactivate',
            'isReversible'
        ],
        properties: []
    },
    Entry: {
        name: 'Entry',
        methods: [
            'closeEntry',
            'openEntry'
        ],
        properties: []
    },
    EntrySensor: {
        name: 'EntrySensor',
        methods: [],
        properties: [
            'entryOpen'
        ]
    },
    DeviceProvider: {
        name: 'DeviceProvider',
        methods: [
            'getDevice',
            'releaseDevice'
        ],
        properties: []
    },
    DeviceDiscovery: {
        name: 'DeviceDiscovery',
        methods: [
            'adoptDevice',
            'discoverDevices'
        ],
        properties: []
    },
    DeviceCreator: {
        name: 'DeviceCreator',
        methods: [
            'createDevice',
            'getCreateDeviceSettings'
        ],
        properties: []
    },
    Battery: {
        name: 'Battery',
        methods: [],
        properties: [
            'batteryLevel'
        ]
    },
    Charger: {
        name: 'Charger',
        methods: [],
        properties: [
            'chargeState'
        ]
    },
    Reboot: {
        name: 'Reboot',
        methods: [
            'reboot'
        ],
        properties: []
    },
    Refresh: {
        name: 'Refresh',
        methods: [
            'getRefreshFrequency',
            'refresh'
        ],
        properties: []
    },
    MediaPlayer: {
        name: 'MediaPlayer',
        methods: [
            'getMediaStatus',
            'load',
            'seek',
            'skipNext',
            'skipPrevious'
        ],
        properties: []
    },
    Online: {
        name: 'Online',
        methods: [],
        properties: [
            'online'
        ]
    },
    BufferConverter: {
        name: 'BufferConverter',
        methods: [
            'convert'
        ],
        properties: [
            'fromMimeType',
            'toMimeType'
        ]
    },
    Settings: {
        name: 'Settings',
        methods: [
            'getSettings',
            'putSetting'
        ],
        properties: []
    },
    BinarySensor: {
        name: 'BinarySensor',
        methods: [],
        properties: [
            'binaryState'
        ]
    },
    TamperSensor: {
        name: 'TamperSensor',
        methods: [],
        properties: [
            'tampered'
        ]
    },
    PowerSensor: {
        name: 'PowerSensor',
        methods: [],
        properties: [
            'powerDetected'
        ]
    },
    AudioSensor: {
        name: 'AudioSensor',
        methods: [],
        properties: [
            'audioDetected'
        ]
    },
    MotionSensor: {
        name: 'MotionSensor',
        methods: [],
        properties: [
            'motionDetected'
        ]
    },
    AmbientLightSensor: {
        name: 'AmbientLightSensor',
        methods: [],
        properties: [
            'ambientLight'
        ]
    },
    OccupancySensor: {
        name: 'OccupancySensor',
        methods: [],
        properties: [
            'occupied'
        ]
    },
    FloodSensor: {
        name: 'FloodSensor',
        methods: [],
        properties: [
            'flooded'
        ]
    },
    UltravioletSensor: {
        name: 'UltravioletSensor',
        methods: [],
        properties: [
            'ultraviolet'
        ]
    },
    LuminanceSensor: {
        name: 'LuminanceSensor',
        methods: [],
        properties: [
            'luminance'
        ]
    },
    PositionSensor: {
        name: 'PositionSensor',
        methods: [],
        properties: [
            'position'
        ]
    },
    SecuritySystem: {
        name: 'SecuritySystem',
        methods: [
            'armSecuritySystem',
            'disarmSecuritySystem'
        ],
        properties: [
            'securitySystemState'
        ]
    },
    PM10Sensor: {
        name: 'PM10Sensor',
        methods: [],
        properties: [
            'pm10Density'
        ]
    },
    PM25Sensor: {
        name: 'PM25Sensor',
        methods: [],
        properties: [
            'pm25Density'
        ]
    },
    VOCSensor: {
        name: 'VOCSensor',
        methods: [],
        properties: [
            'vocDensity'
        ]
    },
    NOXSensor: {
        name: 'NOXSensor',
        methods: [],
        properties: [
            'noxDensity'
        ]
    },
    CO2Sensor: {
        name: 'CO2Sensor',
        methods: [],
        properties: [
            'co2ppm'
        ]
    },
    AirQualitySensor: {
        name: 'AirQualitySensor',
        methods: [],
        properties: [
            'airQuality'
        ]
    },
    AirPurifier: {
        name: 'AirPurifier',
        methods: [
            'setAirPurifierState'
        ],
        properties: [
            'airPurifierState'
        ]
    },
    FilterMaintenance: {
        name: 'FilterMaintenance',
        methods: [],
        properties: [
            'filterChangeIndication',
            'filterLifeLevel'
        ]
    },
    Readme: {
        name: 'Readme',
        methods: [
            'getReadmeMarkdown'
        ],
        properties: []
    },
    OauthClient: {
        name: 'OauthClient',
        methods: [
            'getOauthUrl',
            'onOauthCallback'
        ],
        properties: []
    },
    MixinProvider: {
        name: 'MixinProvider',
        methods: [
            'canMixin',
            'getMixin',
            'releaseMixin'
        ],
        properties: []
    },
    HttpRequestHandler: {
        name: 'HttpRequestHandler',
        methods: [
            'onRequest'
        ],
        properties: []
    },
    EngineIOHandler: {
        name: 'EngineIOHandler',
        methods: [
            'onConnection'
        ],
        properties: []
    },
    PushHandler: {
        name: 'PushHandler',
        methods: [
            'onPush'
        ],
        properties: []
    },
    Program: {
        name: 'Program',
        methods: [
            'run'
        ],
        properties: []
    },
    Scriptable: {
        name: 'Scriptable',
        methods: [
            'eval',
            'loadScripts',
            'saveScript'
        ],
        properties: []
    },
    ObjectTracker: {
        name: 'ObjectTracker',
        methods: [
            'trackObjects'
        ],
        properties: []
    },
    ObjectDetector: {
        name: 'ObjectDetector',
        methods: [
            'getDetectionInput',
            'getObjectTypes'
        ],
        properties: []
    },
    ObjectDetection: {
        name: 'ObjectDetection',
        methods: [
            'detectObjects',
            'generateObjectDetections',
            'getDetectionModel'
        ],
        properties: []
    },
    ObjectDetectionPreview: {
        name: 'ObjectDetectionPreview',
        methods: [],
        properties: []
    },
    ObjectDetectionGenerator: {
        name: 'ObjectDetectionGenerator',
        methods: [],
        properties: []
    },
    HumiditySetting: {
        name: 'HumiditySetting',
        methods: [
            'setHumidity'
        ],
        properties: [
            'humiditySetting'
        ]
    },
    Fan: {
        name: 'Fan',
        methods: [
            'setFan'
        ],
        properties: [
            'fan'
        ]
    },
    RTCSignalingChannel: {
        name: 'RTCSignalingChannel',
        methods: [
            'startRTCSignalingSession'
        ],
        properties: []
    },
    RTCSignalingClient: {
        name: 'RTCSignalingClient',
        methods: [
            'createRTCSignalingSession'
        ],
        properties: []
    },
    LauncherApplication: {
        name: 'LauncherApplication',
        methods: [],
        properties: [
            'applicationInfo'
        ]
    },
    ScryptedUser: {
        name: 'ScryptedUser',
        methods: [
            'getScryptedUserAccessControl'
        ],
        properties: []
    },
    VideoFrameGenerator: {
        name: 'VideoFrameGenerator',
        methods: [
            'generateVideoFrames'
        ],
        properties: []
    },
    StreamService: {
        name: 'StreamService',
        methods: [
            'connectStream'
        ],
        properties: []
    }
};
/**
 * @category Core Reference
 */
var ScryptedDeviceType;
(function (ScryptedDeviceType) {
    ScryptedDeviceType["Builtin"] = "Builtin";
    ScryptedDeviceType["Camera"] = "Camera";
    ScryptedDeviceType["Fan"] = "Fan";
    ScryptedDeviceType["Light"] = "Light";
    ScryptedDeviceType["Switch"] = "Switch";
    ScryptedDeviceType["Outlet"] = "Outlet";
    ScryptedDeviceType["Sensor"] = "Sensor";
    ScryptedDeviceType["Scene"] = "Scene";
    ScryptedDeviceType["Program"] = "Program";
    ScryptedDeviceType["Automation"] = "Automation";
    ScryptedDeviceType["Vacuum"] = "Vacuum";
    ScryptedDeviceType["Notifier"] = "Notifier";
    ScryptedDeviceType["Thermostat"] = "Thermostat";
    ScryptedDeviceType["Lock"] = "Lock";
    ScryptedDeviceType["PasswordControl"] = "PasswordControl";
    /**
     * Displays have audio and video output.
     */
    ScryptedDeviceType["Display"] = "Display";
    /**
     * Smart Displays have two way audio and video.
     */
    ScryptedDeviceType["SmartDisplay"] = "SmartDisplay";
    ScryptedDeviceType["Speaker"] = "Speaker";
    /**
     * Smart Speakers have two way audio.
     */
    ScryptedDeviceType["SmartSpeaker"] = "SmartSpeaker";
    ScryptedDeviceType["Event"] = "Event";
    ScryptedDeviceType["Entry"] = "Entry";
    ScryptedDeviceType["Garage"] = "Garage";
    ScryptedDeviceType["DeviceProvider"] = "DeviceProvider";
    ScryptedDeviceType["DataSource"] = "DataSource";
    ScryptedDeviceType["API"] = "API";
    ScryptedDeviceType["Doorbell"] = "Doorbell";
    ScryptedDeviceType["Irrigation"] = "Irrigation";
    ScryptedDeviceType["Valve"] = "Valve";
    ScryptedDeviceType["Person"] = "Person";
    ScryptedDeviceType["SecuritySystem"] = "SecuritySystem";
    ScryptedDeviceType["WindowCovering"] = "WindowCovering";
    ScryptedDeviceType["Siren"] = "Siren";
    ScryptedDeviceType["AirPurifier"] = "AirPurifier";
    ScryptedDeviceType["Unknown"] = "Unknown";
})(ScryptedDeviceType = exports.ScryptedDeviceType || (exports.ScryptedDeviceType = {}));
var HumidityMode;
(function (HumidityMode) {
    HumidityMode["Humidify"] = "Humidify";
    HumidityMode["Dehumidify"] = "Dehumidify";
    HumidityMode["Auto"] = "Auto";
    HumidityMode["Off"] = "Off";
})(HumidityMode = exports.HumidityMode || (exports.HumidityMode = {}));
var FanMode;
(function (FanMode) {
    FanMode["Auto"] = "Auto";
    FanMode["Manual"] = "Manual";
})(FanMode = exports.FanMode || (exports.FanMode = {}));
var TemperatureUnit;
(function (TemperatureUnit) {
    TemperatureUnit["C"] = "C";
    TemperatureUnit["F"] = "F";
})(TemperatureUnit = exports.TemperatureUnit || (exports.TemperatureUnit = {}));
var ThermostatMode;
(function (ThermostatMode) {
    ThermostatMode["Off"] = "Off";
    ThermostatMode["Cool"] = "Cool";
    ThermostatMode["Heat"] = "Heat";
    ThermostatMode["HeatCool"] = "HeatCool";
    ThermostatMode["Auto"] = "Auto";
    ThermostatMode["FanOnly"] = "FanOnly";
    ThermostatMode["Purifier"] = "Purifier";
    ThermostatMode["Eco"] = "Eco";
    ThermostatMode["Dry"] = "Dry";
    ThermostatMode["On"] = "On";
})(ThermostatMode = exports.ThermostatMode || (exports.ThermostatMode = {}));
var PanTiltZoomMovement;
(function (PanTiltZoomMovement) {
    PanTiltZoomMovement["Absolute"] = "Absolute";
    PanTiltZoomMovement["Relative"] = "Relative";
})(PanTiltZoomMovement = exports.PanTiltZoomMovement || (exports.PanTiltZoomMovement = {}));
var LockState;
(function (LockState) {
    LockState["Locked"] = "Locked";
    LockState["Unlocked"] = "Unlocked";
    LockState["Jammed"] = "Jammed";
})(LockState = exports.LockState || (exports.LockState = {}));
var ChargeState;
(function (ChargeState) {
    ChargeState["Trickle"] = "trickle";
    ChargeState["Charging"] = "charging";
    ChargeState["NotCharging"] = "not-charging";
})(ChargeState = exports.ChargeState || (exports.ChargeState = {}));
var AirPurifierStatus;
(function (AirPurifierStatus) {
    AirPurifierStatus["Inactive"] = "Inactive";
    AirPurifierStatus["Idle"] = "Idle";
    AirPurifierStatus["Active"] = "Active";
    AirPurifierStatus["ActiveNightMode"] = "ActiveNightMode";
})(AirPurifierStatus = exports.AirPurifierStatus || (exports.AirPurifierStatus = {}));
var AirPurifierMode;
(function (AirPurifierMode) {
    AirPurifierMode["Manual"] = "Manual";
    AirPurifierMode["Automatic"] = "Automatic";
})(AirPurifierMode = exports.AirPurifierMode || (exports.AirPurifierMode = {}));
var AirQuality;
(function (AirQuality) {
    AirQuality["Unknown"] = "Unknown";
    AirQuality["Excellent"] = "Excellent";
    AirQuality["Good"] = "Good";
    AirQuality["Fair"] = "Fair";
    AirQuality["Inferior"] = "Inferior";
    AirQuality["Poor"] = "Poor";
})(AirQuality = exports.AirQuality || (exports.AirQuality = {}));
var SecuritySystemMode;
(function (SecuritySystemMode) {
    SecuritySystemMode["Disarmed"] = "Disarmed";
    SecuritySystemMode["HomeArmed"] = "HomeArmed";
    SecuritySystemMode["AwayArmed"] = "AwayArmed";
    SecuritySystemMode["NightArmed"] = "NightArmed";
})(SecuritySystemMode = exports.SecuritySystemMode || (exports.SecuritySystemMode = {}));
var SecuritySystemObstruction;
(function (SecuritySystemObstruction) {
    SecuritySystemObstruction["Sensor"] = "Sensor";
    SecuritySystemObstruction["Occupied"] = "Occupied";
    SecuritySystemObstruction["Time"] = "Time";
    SecuritySystemObstruction["Error"] = "Error";
})(SecuritySystemObstruction = exports.SecuritySystemObstruction || (exports.SecuritySystemObstruction = {}));
var MediaPlayerState;
(function (MediaPlayerState) {
    MediaPlayerState["Idle"] = "Idle";
    MediaPlayerState["Playing"] = "Playing";
    MediaPlayerState["Paused"] = "Paused";
    MediaPlayerState["Buffering"] = "Buffering";
})(MediaPlayerState = exports.MediaPlayerState || (exports.MediaPlayerState = {}));
var ScryptedInterface;
(function (ScryptedInterface) {
    ScryptedInterface["ScryptedDevice"] = "ScryptedDevice";
    ScryptedInterface["ScryptedPlugin"] = "ScryptedPlugin";
    ScryptedInterface["OnOff"] = "OnOff";
    ScryptedInterface["Brightness"] = "Brightness";
    ScryptedInterface["ColorSettingTemperature"] = "ColorSettingTemperature";
    ScryptedInterface["ColorSettingRgb"] = "ColorSettingRgb";
    ScryptedInterface["ColorSettingHsv"] = "ColorSettingHsv";
    ScryptedInterface["Notifier"] = "Notifier";
    ScryptedInterface["StartStop"] = "StartStop";
    ScryptedInterface["Pause"] = "Pause";
    ScryptedInterface["Dock"] = "Dock";
    ScryptedInterface["TemperatureSetting"] = "TemperatureSetting";
    ScryptedInterface["Thermometer"] = "Thermometer";
    ScryptedInterface["HumiditySensor"] = "HumiditySensor";
    ScryptedInterface["Camera"] = "Camera";
    ScryptedInterface["Microphone"] = "Microphone";
    ScryptedInterface["Display"] = "Display";
    ScryptedInterface["VideoCamera"] = "VideoCamera";
    ScryptedInterface["VideoRecorder"] = "VideoRecorder";
    ScryptedInterface["VideoRecorderManagement"] = "VideoRecorderManagement";
    ScryptedInterface["PanTiltZoom"] = "PanTiltZoom";
    ScryptedInterface["EventRecorder"] = "EventRecorder";
    ScryptedInterface["VideoClips"] = "VideoClips";
    ScryptedInterface["VideoCameraConfiguration"] = "VideoCameraConfiguration";
    ScryptedInterface["Intercom"] = "Intercom";
    ScryptedInterface["Lock"] = "Lock";
    ScryptedInterface["PasswordStore"] = "PasswordStore";
    ScryptedInterface["Scene"] = "Scene";
    ScryptedInterface["Entry"] = "Entry";
    ScryptedInterface["EntrySensor"] = "EntrySensor";
    ScryptedInterface["DeviceProvider"] = "DeviceProvider";
    ScryptedInterface["DeviceDiscovery"] = "DeviceDiscovery";
    ScryptedInterface["DeviceCreator"] = "DeviceCreator";
    ScryptedInterface["Battery"] = "Battery";
    ScryptedInterface["Charger"] = "Charger";
    ScryptedInterface["Reboot"] = "Reboot";
    ScryptedInterface["Refresh"] = "Refresh";
    ScryptedInterface["MediaPlayer"] = "MediaPlayer";
    ScryptedInterface["Online"] = "Online";
    ScryptedInterface["BufferConverter"] = "BufferConverter";
    ScryptedInterface["Settings"] = "Settings";
    ScryptedInterface["BinarySensor"] = "BinarySensor";
    ScryptedInterface["TamperSensor"] = "TamperSensor";
    ScryptedInterface["PowerSensor"] = "PowerSensor";
    ScryptedInterface["AudioSensor"] = "AudioSensor";
    ScryptedInterface["MotionSensor"] = "MotionSensor";
    ScryptedInterface["AmbientLightSensor"] = "AmbientLightSensor";
    ScryptedInterface["OccupancySensor"] = "OccupancySensor";
    ScryptedInterface["FloodSensor"] = "FloodSensor";
    ScryptedInterface["UltravioletSensor"] = "UltravioletSensor";
    ScryptedInterface["LuminanceSensor"] = "LuminanceSensor";
    ScryptedInterface["PositionSensor"] = "PositionSensor";
    ScryptedInterface["SecuritySystem"] = "SecuritySystem";
    ScryptedInterface["PM10Sensor"] = "PM10Sensor";
    ScryptedInterface["PM25Sensor"] = "PM25Sensor";
    ScryptedInterface["VOCSensor"] = "VOCSensor";
    ScryptedInterface["NOXSensor"] = "NOXSensor";
    ScryptedInterface["CO2Sensor"] = "CO2Sensor";
    ScryptedInterface["AirQualitySensor"] = "AirQualitySensor";
    ScryptedInterface["AirPurifier"] = "AirPurifier";
    ScryptedInterface["FilterMaintenance"] = "FilterMaintenance";
    ScryptedInterface["Readme"] = "Readme";
    ScryptedInterface["OauthClient"] = "OauthClient";
    ScryptedInterface["MixinProvider"] = "MixinProvider";
    ScryptedInterface["HttpRequestHandler"] = "HttpRequestHandler";
    ScryptedInterface["EngineIOHandler"] = "EngineIOHandler";
    ScryptedInterface["PushHandler"] = "PushHandler";
    ScryptedInterface["Program"] = "Program";
    ScryptedInterface["Scriptable"] = "Scriptable";
    ScryptedInterface["ObjectTracker"] = "ObjectTracker";
    ScryptedInterface["ObjectDetector"] = "ObjectDetector";
    ScryptedInterface["ObjectDetection"] = "ObjectDetection";
    ScryptedInterface["ObjectDetectionPreview"] = "ObjectDetectionPreview";
    ScryptedInterface["ObjectDetectionGenerator"] = "ObjectDetectionGenerator";
    ScryptedInterface["HumiditySetting"] = "HumiditySetting";
    ScryptedInterface["Fan"] = "Fan";
    ScryptedInterface["RTCSignalingChannel"] = "RTCSignalingChannel";
    ScryptedInterface["RTCSignalingClient"] = "RTCSignalingClient";
    ScryptedInterface["LauncherApplication"] = "LauncherApplication";
    ScryptedInterface["ScryptedUser"] = "ScryptedUser";
    ScryptedInterface["VideoFrameGenerator"] = "VideoFrameGenerator";
    ScryptedInterface["StreamService"] = "StreamService";
})(ScryptedInterface = exports.ScryptedInterface || (exports.ScryptedInterface = {}));
var ScryptedMimeTypes;
(function (ScryptedMimeTypes) {
    ScryptedMimeTypes["Url"] = "text/x-uri";
    ScryptedMimeTypes["InsecureLocalUrl"] = "text/x-insecure-local-uri";
    ScryptedMimeTypes["LocalUrl"] = "text/x-local-uri";
    ScryptedMimeTypes["PushEndpoint"] = "text/x-push-endpoint";
    ScryptedMimeTypes["SchemePrefix"] = "x-scrypted/x-scrypted-scheme-";
    ScryptedMimeTypes["MediaStreamUrl"] = "text/x-media-url";
    ScryptedMimeTypes["MediaObject"] = "x-scrypted/x-scrypted-media-object";
    ScryptedMimeTypes["RequestMediaObject"] = "x-scrypted/x-scrypted-request-media-object";
    ScryptedMimeTypes["RequestMediaStream"] = "x-scrypted/x-scrypted-request-stream";
    ScryptedMimeTypes["MediaStreamFeedback"] = "x-scrypted/x-media-stream-feedback";
    ScryptedMimeTypes["FFmpegInput"] = "x-scrypted/x-ffmpeg-input";
    ScryptedMimeTypes["FFmpegTranscodeStream"] = "x-scrypted/x-ffmpeg-transcode-stream";
    ScryptedMimeTypes["RTCSignalingChannel"] = "x-scrypted/x-scrypted-rtc-signaling-channel";
    ScryptedMimeTypes["RTCSignalingSession"] = "x-scrypted/x-scrypted-rtc-signaling-session";
    ScryptedMimeTypes["RTCConnectionManagement"] = "x-scrypted/x-scrypted-rtc-connection-management";
    ScryptedMimeTypes["Image"] = "x-scrypted/x-scrypted-image";
})(ScryptedMimeTypes = exports.ScryptedMimeTypes || (exports.ScryptedMimeTypes = {}));
